tests/test_page.py

```diff
@@ -10,23 +10,48 @@
 
 
 def test_base_url(base_url, page):
-    assert base_url == page.url
+    assert base_url == page.canonical_url
 
 
-def test_url_keywords(base_url, selenium):
+def test_canonical_url_absolute(base_url, selenium):
+    url_template = 'https://www.test.com/'
+
+    class MyPage(Page):
+        URL_TEMPLATE = url_template
+    page = MyPage(selenium, base_url)
+    assert url_template == page.canonical_url
+
+
+def test_canonical_url_empty(selenium):
+    page = Page(selenium)
+    assert page.canonical_url == ''
+
+
+def test_canonical_url_keywords(base_url, selenium):
     value = str(random.random())
 
     class MyPage(Page):
-        _url = '{key}'
-    page = MyPage('foo', selenium, key=value)
-    assert value == page.url
+        URL_TEMPLATE = '{key}'
+    page = MyPage(selenium, base_url, key=value)
+    assert base_url + value == page.canonical_url
+
+
+def test_canonical_url_prepend(base_url, page, selenium):
+    assert base_url == page.canonical_url
```

Comment:
I disagree that it isn't testing prepending. If prepending wasn't working then we wouldn't get the `base_url` back from `canonical_url`. Although I agree that it theoretically could be appending, or replacing, so something that uses a `URL_TEMPLATE` would be better. Having said that, `test_canonical_url_keywords` already does that, I believe. It uses a `URL_TEMPLATE` and a `base_url`, so I think we already have the test you mention. Perhaps I should just remove this test and we can rely on `test_canonical_url_keywords` to test both base_url prepending and keyword replacement? What do you think?